tensorflow/python/tools/optimize_for_inference_lib.py

```diff
@@ -728,3 +731,265 @@ def convert_placeholder_to_const(input_graph_def, nodes_to_convert=None):
+               {"Mul"  // same mul_1 op as in previous block
+                  {{"Rsqrt"
+                      {"Add"
+                          {{"Const"},  // Variance
+                           {"Const"}  // Epsilon
+                          }
+                      }
+                   },  // end - Rsqrt
+                   {"Const"}  // Gamma
+                  }
+                }  // end - mul_1
+              }
+            }  // end - mul_3
+          }
+        }  // end - Sub
+      }
+  }  // end - Add
+
+  Subgraph pattern when gamma value is 1 and the gamma scaling Mul is skipped
+  {"Add"
+      {{"Mul"  // mul_0
+          {{"*"},  // input to apply batchnorma
+           {"Rsqrt"  // same Rsqrt op used in Sub block
+              {"Add"
+                 {{"Const"},  // Variance
+                  {"Const"}  // Epsilon
+                 }
+              }
+            }  // end - Rsqrt
+          }
+        },  // end - mul_0
+        {"Sub"
+          {{"Const"},  // Beta
+           {"Mul"  // mul_1
+              {{"Const"},  // Mean
+               {"Rsqrt"  // same Rsqrt op as in previous mul_0 block
+                  {"Add"
+                    {{"Const"},  // Variance
+                     {"Const"}  // Epsilon
+                    }
+                  }
+                }  // end - Rsqrt
+              }
+           }  // end - mul_1
+          }
+        }  // end - Sub
+      }
+  }  // end - Add
+
+  Args:
+    input_graph_def: A GraphDef containing a model.
+
+  Returns:
+    Modified graph with individual ops that made up of batch normalization
+    fused to FusedBatchNorm.
+
+  Raises:
+    ValueError: If the graph is badly formed with duplicate node names.
+  """
+  input_node_map = {}
+  for node in input_graph_def.node:
+    if node.name not in input_node_map:
+      input_node_map[node.name] = node
+    else:
+      raise ValueError("Duplicate node names detected for ", node.name)
+
+  nodes_to_skip = {}
+  new_ops = []
+  for node in input_graph_def.node:
+    if node.op != "Add":
+      continue
+
+    # Add (Mul, Sub) or Add (Sub, Mul)
+    input0_op = node_from_map(input_node_map, node.input[0])
+    input1_op = node_from_map(input_node_map, node.input[1])
+
+    if input0_op.op == "Mul" and input1_op.op == "Sub":
+      data_scale_mul_op = input0_op
+      bias_mean_sub_op = input1_op
+    elif input0_op.op == "Sub" and input1_op.op == "Mul":
+      bias_mean_sub_op = input0_op
+      data_scale_mul_op = input1_op
+    else:
+      continue
+
+    # Mul (input, Mul)
+    input_data_op = node_from_map(input_node_map, data_scale_mul_op.input[0])
+    scale_op = node_from_map(input_node_map, data_scale_mul_op.input[1])
+
+    # Check input to batchnorm and only proceed fusion if input is
+    # Conv2D or DepthwiseConv2dNative and data format is NHWC.
+    data_format = None
+    if input_data_op.op in ["Conv2D", "DepthwiseConv2dNative"]:
+      data_format = input_data_op.attr["data_format"]
```

Comment:
Our tools say this isn't covered by tests.

Reply:
I added unit test case for the data format.